src/nb_foreign.cpp

```diff
@@ -0,0 +1,541 @@
+    fw->translate_exception = nb_foreign_translate_exception;
+    fw->add_foreign_binding = nb_foreign_add_foreign_binding;
+    fw->remove_foreign_binding = nb_foreign_remove_foreign_binding;
+    fw->add_foreign_framework = nb_foreign_add_foreign_framework;
+    internals_->foreign_self = fw;
+
+    auto *registry = internals_->foreign_registry;
+    // pymb_add_framework() will call our add_foreign_framework and
+    // add_foreign_binding method for each existing other framework/binding;
+    // those need to lock internals, so unlock here
+    unlock_internals guard{internals_};
+    pymb_add_framework(registry, fw);
+}
+
+// Add the given `binding` to our type maps so that we can use it to satisfy
+// from- and to-Python requests for the given C++ type
+static void nb_type_import_binding(pymb_binding *binding,
+                                   const std::type_info *cpptype) noexcept {
+    // Caller must hold the internals lock
+    internals->foreign_imported_any = true;
+
+    auto add_to_list = [binding](void *list_head) -> void* {
+        if (!list_head) {
+            return binding;
+        }
+        nb_foreign_seq *seq = nb_ensure_seq<pymb_binding>(&list_head);
+        while (true) {
+            if (seq->value == binding)
+                return list_head; // already added
+            nb_foreign_seq *next = nb_load_acquire(seq->next);
+            if (next == nullptr)
+                break;
+            seq = next;
+        }
+        nb_foreign_seq *next =
+                (nb_foreign_seq *) PyMem_Malloc(sizeof(nb_foreign_seq));
+        check(next, "add_foreign_binding_to_list(): out of memory!");
+        next->value = binding;
+        next->next = nullptr;
+        nb_store_release(seq->next, next);
+        return list_head;
+    };
+
+    auto [it, inserted] = internals->type_c2p_slow.try_emplace(
+             cpptype, nb_mark_foreign(binding));
+    if (!inserted) {
+        if (nb_is_foreign(it->second))
+            it.value() = nb_mark_foreign(add_to_list(nb_get_foreign(it->second)));
+        else if (auto *t = (type_data *) it->second)
+            nb_store_release(t->foreign_bindings,
+                             add_to_list(nb_load_acquire(t->foreign_bindings)));
+        else
+            check(false, "null entry in type_c2p_slow");
+    }
+    nb_type_update_c2p_fast(cpptype, it->second);
+}
+
+// Learn to satisfy from- and to-Python requests for `cpptype` using the
+// foreign binding provided by the given `pytype`. If cpptype is nullptr, infer
+// the C++ type by looking at the binding, and require that its ABI match ours.
+// Throws an exception on failure. Caller must hold the internals lock.
+void nb_type_import_impl(PyObject *pytype, const std::type_info *cpptype) {
+    if (!internals->foreign_registry)
+        register_with_pymetabind(internals);
+    pymb_framework* foreign_self = internals->foreign_self;
+    pymb_binding* binding = pymb_get_binding(pytype);
```

Comment:
So this is a naïve question. Why would we need to look up ``__pymetabind_binding__``? Won't we be notified of new frameworks/bindings using the hooks?

Or is the idea that the metabind feature is enabled lazily, and if we join the party late then that registry is empty to start with? Still then, I am wondering why we can't populate our own tables from the list of types in the registry, without having to touch the ``__pymetabind_binding__`` member.